binding.c

```diff
@@ -231,40 +309,45 @@ on_udx_teardown (napi_async_cleanup_hook_handle handle, void *data) {
 }
 
 static void
-ensure_teardown (napi_env env, udx_napi_t *udx) {
+ensure_teardown (js_env_t *env, udx_napi_t *udx) {
   if (udx->has_teardown) return;
   udx->has_teardown = true;
-  napi_add_async_cleanup_hook(env, on_udx_teardown, (void *) udx, &(udx->teardown));
+
+  int err = js_add_deferred_teardown_callback(env, on_udx_teardown, (void *) udx, &(udx->teardown));
+  if (err != 0) abort();
 }
 
 static void
 on_udx_stream_end (udx_stream_t *stream) {
+  int err;
   udx_napi_stream_t *n = (udx_napi_stream_t *) stream;
   if (n->udx->exiting) return;
 
   size_t read = n->read_buf_head - n->read_buf;
 
-  napi_env env = n->env;
+  js_env_t *env = n->env;
 
-  napi_handle_scope scope;
-  napi_open_handle_scope(env, &scope);
+  js_handle_scope_t *scope;
+  err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
 
-  napi_value ctx;
-  napi_get_reference_value(env, n->ctx, &ctx);
+  js_value_t *ctx;
+  err = js_get_reference_value(env, n->ctx, &ctx);
+  assert(err == 0);
 
-  napi_value callback;
-  napi_get_reference_value(env, n->on_end, &callback);
+  js_value_t *callback;
+  err = js_get_reference_value(env, n->on_end, &callback);
+  assert(err == 0);
 
-  napi_value argv[1];
-  napi_create_uint32(env, read, &(argv[0]));
+  js_value_t *argv[1];
+  err = js_create_uint32(env, read, &(argv[0]));
+  assert(err == 0);
 
-  if (napi_make_callback(env, NULL, ctx, callback, 1, argv, NULL) == napi_pending_exception) {
-    napi_value fatal_exception;
-    napi_get_and_clear_last_exception(env, &fatal_exception);
-    napi_fatal_exception(env, fatal_exception);
-  }
+  err = js_call_function(env, ctx, callback, 1, argv, NULL);
```

Comment:
This should do a microtask checkpoint assuming that `on_udx_stream_end()` is an I/O callback.